source/agora/consensus/state/UTXOSet.d

```diff
@@ -137,59 +117,214 @@ public class UTXOSet
 
         Returns:
-            the associative array for UTXOs
+            the number of elements in the UTXO set
 
     ***************************************************************************/
 
-    public UTXO[Hash] getUTXOs (const ref PublicKey pubkey) @safe nothrow
+    public size_t length () @safe
     {
-        return this.utxo_db.getUTXOs(pubkey);
+        return this.utxo_db.length();
     }
 
     /***************************************************************************
 
-        Find an UTXO in the UTXO set.
+        Get UTXOs from the UTXO set
 
         Params:
-            hash = the hash of the UTXO (`hashFull(tx_hash, index)`)
-            output = will contain the UTXO if found
+            pubkey = the key by which to search UTXOs in UTXOSet
 
-        Return:
-            Return true if the UTXO was found
+        Returns:
+            the associative array for UTXOs
 
     ***************************************************************************/
 
-    private bool findUTXO (Hash utxo, out UTXO value)
-        nothrow @safe
+    public UTXO[Hash] getUTXOs (const ref PublicKey pubkey) nothrow @safe
     {
-        if (utxo in this.used_utxos)
+        return this.utxo_db.getUTXOs(pubkey);
+    }
+
+    ///
+    public override bool peekUTXO (Hash utxo, out UTXO value) nothrow @safe
+    {
+        return this.utxo_db.find(utxo, value);
+    }
+
+    ///
+    protected override void remove (Hash utxo) @safe
+    {
+        this.utxo_db.remove(utxo);
+    }
+
+    ///
+    protected override void add (Hash utxo, UTXO value) @safe
+    {
+        this.utxo_db[utxo] = value;
+    }
+}
+
+/*******************************************************************************
+
+    This is a simple UTXOSet, used when the AA behavior is desired
+
+    Most unittests do not need a fully-fledged UTXOSet with all the DB and
+    serialization that comes with it, instead relying on an associative array
+    and a delegate.
+
+    Since this pattern is so common, this class is offered as a mean to achieve
+    this without code duplication. See issue #501 for history.
+
+    Note that this should *NOT* be used to replace the above UTXOSet,
+    when for example doing integration tests with LocalRest.
+
+*******************************************************************************/
+
+public class TestUTXOSet : UTXOCache
+{
+    /// UTXO cache backed by an AA
+    public UTXO[Hash] storage;
+
+    ///
+    alias storage this;
+
+    /// Short hand to add a transaction
+    public void put (const Transaction tx) nothrow @safe
+    {
+        Hash txhash = hashFull(tx);
+        foreach (size_t idx, ref output_; tx.outputs)
         {
-            log.trace("findUTXO: utxo_hash {} found in used_utxos: {}", utxo, used_utxos);
-            return false;  // double-spend
+            Hash h = UTXO.getHash(txhash, idx);
+            UTXO v = {
+                type: tx.type,
```

Comment:
Something I have noticed with the `TestUTXOSet` is that it doesn't store the `unlock_height`. Could we add it as a parameter which defaults to 0?

Reply:
Good point. Although we now have `updateUTXOCache` through the base class which would do that for you, so I think we could just remove this method in the long run ?

Reply:
Ah right. Yes then it makes sense to just remove it, otherwise it's duplicating the functionality (poorly).

Reply:
Correct, `updateUTXOCache` can now be called with `TestUTXOSet`. I was thinking of some uses-cases where we would want to "stack" the UTXOs (not remove any).
If there isn't such a case, then we don't need it.

Reply:
I tried to replace occurrences of `put` with `updateUTXOCache`, but it doesn't seem to be a simple replace.
Could it be the fact that `put` doesn't have `unlock_height`
```D
UTXO v = {
    type: tx.type,
    output: output_
};
```
and `updateUTXOCache` takes into account the `unlock_height`?

Reply:
Setting it to `0` should work then, no?